### PR TITLE
feat(dlx): cache

### PR DESCRIPTION
**Pull request by KSXGitHub**
Fixes https://github.com/pnpm/pnpm/issues/5277

---

**TODO:**
* [x] Group the cache directories by id.
* [x] Fix tests in `dlx.e2e.ts` and `run.ts`.
* [x] Fix `cleanExpiredCache.ts`.
* [x] Fix `cleanExpiredCache.test.ts`.
* [x] Refactor: Rename `cleanExpiredCache` to `cleanExpiredDlxCache`.
* [x] Fix tests for `storePrune.ts`.
* [x] Proper tests for pruning dlx cache must be placed in `run.ts` (as a CLI invoker).
* [x] Test removing orphans.
* [x] Test running `dlx` with cache that is expired but still exist.
* [x] Test running multiple `dlx` in parallel with cache that is expired but still exist.
* [x] Refactor `dlx.ts`: async fs methods.

### CODE REVIEW

**Comment by zkochan** [comment]
The way it is implemented, the cache max age setting is just ignored because the cached directory is always used by dlx. So it is basically an ageless cache. I don't think it is a good idea to implement it like this.

We need to check the age of the cache. If it is too old, pnpm shouldn't use it and it should create a new cache. Of course, it should be atomic. The current approach with mkdir and checking for existence of a cache directory will not work in this case (because we will also remove it sometimes, so there won't be a way to make it atomic between several running processes of pnpm). We can do a different approach. pnpm will install to a "temp" location but instead of a cached directory there will be a cached symlink (junction on Windows) pointing to the "temp" location (so actually it won't be temp directory, just a location with a unique random name).

```
symlink            here pnpm runs install
<cached dir> ----> <cached dir>_<process id>
```

before the installation the symlink should be removed and only after a successful install should the symlink be created. This will guarantee that the cache won't ever be broken.

**Comment by KSXGitHub** [comment]
Actually, is symlink even necessary at all? We can install all the directories to the `cache-dir` without creating any symlink. On each `dlx` call, we can check for a fully installed directory (by checking the existence of `$DIR/node_modules/.modules.yaml`) that is up-to-date and use that if exist. If none of the cached directories satisfy the requirements, simply create a new one. This would at least simplify the code.

**Comment by zkochan** [comment]
I don't understand how your idea would work. If the cache doesn't exist and the same dlx command will be executed simultaneously, the cached directory can become corrupted.

**Comment by KSXGitHub** [comment]
> If the cache doesn't exist and the same dlx command will be executed simultaneously

Multiple new caches shall be created. The caches are guaranteed to all have different names because they have date + pid suffix.

**Comment by KSXGitHub** [comment]
The tests in `dlx.e2e.ts` and `run.ts` should pass. But not the tests in `storePrune.ts` and `cleanExpiredCache.test.ts` because I haven't touched them.

**Comment by zkochan** [comment]
So you will search for multiple cache directories and pick one of them? Still, creating a symlink on success is a better approach than checking the existence of `node_modules/.modules.yaml`.

**Comment by KSXGitHub** [comment]
> Still, creating a symlink on success is a better approach than checking the existence of `node_modules/.modules.yaml`.

It would have to check if the symlink is up-to-date. And if it was out-of-date, it would have to unlink then recreate the symlink (because Node.js doesn't provide API to atomically modify a symlink as far as I have looked up).

If multiple `dlx` calls to the same package was executed simultaneously, we would have multiple overlapping `fs.unlinkSync` and `fs.symlinkSync` calls.

**Comment by zkochan** [comment]
I don't see a problem. The symlink will only be created when the cache is valid. There is no scenario when the symlink gets broken or links to broken cache.

File: exec/plugin-commands-script-runners/src/dlx.ts [hunk 1]
```diff
@@ -63,44 +65,62 @@ export function help () {
...
+    }, pkgs)
+    try {
+      await fs.unlink(cacheLink)
+    } catch (err) {
+      if (!(util.types.isNativeError(err) && 'code' in err && err.code === 'ENOENT')) {
+        throw err
+      }
+    }
+    try {
+      await fs.symlink(newPrefix, cacheLink, 'junction')
```
Comment (zkochan) [inline exec/plugin-commands-script-runners/src/dlx.ts]: Use the `symlink-dir` library. It is already used in many places for creating symlinks on POSIX and junctions on Windows.

**Review by zkochan — COMMENTED** [review]
I think overall it looks fine. I will spend more time reviewing it later because it is a lot of changes. However, I think the approach is fine.

File: exec/plugin-commands-script-runners/src/dlx.ts (hunk 1 above)
Comment (KSXGitHub) [reply exec/plugin-commands-script-runners/src/dlx.ts]: https://github.com/pnpm/pnpm/pull/7835/commits/746a079dd3015c341c7b805f3b317dea5d226192

**Comment by zkochan** [comment]
tests fail after your change

**Comment by zkochan** [comment]
The test fails on Windows with Node.js 21 due to some issue released in node.js v21.5.0. This might fix the issue as it seems to happen only when the path is long: https://github.com/nodejs/node/pull/51097

File: .changeset/thin-rice-promise.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+---
+"@pnpm/plugin-commands-script-runners": minor
+"@pnpm/config": minor
+"pnpm": minor
+"@pnpm/plugin-commands-outdated": patch
+"@pnpm/plugin-commands-store": patch
+---
+
+Added cache for `pnpm dlx` [#5277](https://github.com/pnpm/pnpm/issues/5277).
```
Comment (styfle) [inline .changeset/thin-rice-promise.md:9]: Where is it cached and for how long?

Comment (KSXGitHub) [reply .changeset/thin-rice-promise.md:9]: 5 minutes by default. Add `dlx-cache-max-age=N` to `.npmrc` to set it to N minutes.

Comment (styfle) [reply .changeset/thin-rice-promise.md:9]: Thanks! The docs should probably be updated here: https://pnpm.io/cli/dlx

Comment (zkochan) [reply .changeset/thin-rice-promise.md:9]: No, I changed it to 1 day.

**Pull request merged**